### PR TITLE
added SudokuMove and SudokuPuzzle class skeletons

### PR DESCRIPTION
**Pull request by MatthewThompson35**

### CODE REVIEW

**Review by bfreder1 — CHANGES_REQUESTED** [review]
Neither class has a package declaration and should be grouped together in a sudoku package in the model package. Outside of that, they function fine as skeletons.

File: code/model/sudoku/SudokuPuzzle.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
+package code.model.sudoku;
+
+import java.util.Stack;
+import code.model.*;
+
+/**
+ * Sudoku puzzle class
+ */
```
Comment (TakenouchiTR) [inline code/model/sudoku/SudokuPuzzle.java:8]: Missing author and version

File: code/model/sudoku/SudokuMove.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
+package code.model.sudoku;
+/**
+ * The soduku move class
+ *
+ */
```
Comment (TakenouchiTR) [inline code/model/sudoku/SudokuMove.java:5]: Missing author and version

File: code/model/sudoku/SudokuPuzzle.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    private Stack<SudokuMove> moveHistory;
+
+    /**
+     * Default Constructor
+     */
+    public SudokuPuzzle() {
+        this.numbers = new int[9][9];
+        this.numberLocks = new boolean[9][9];
+        this.defaultAnswer = new int[9][9];
+        this.selectedNumber = null;
```
Comment (TakenouchiTR) [inline code/model/sudoku/SudokuPuzzle.java:24]: Unless I'm missing something, you shouldn't be able to set an `int` to `null` in Java.
Recommend using a default value, such as 0 or -1.

File: code/model/sudoku/SudokuPuzzle.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    /***
+     * gets the answer for the specified position
+     * 
+     * @param column
+     * @param row
+     * @return the answer for the coord
+     */
+    public int getAnswerForPosition(int column, int row) {
+        return null;
```
Comment (TakenouchiTR) [inline code/model/sudoku/SudokuPuzzle.java:114]: Method should return that answer for a specified `column` and `row`, not return null.

File: code/model/sudoku/SudokuPuzzle.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    /***
+     * checks to see if number is locked
+     * 
+     * @param column
+     * @param row
+     * @return true or false depending on locked state
+     */
+    public boolean isNumberLocked(int column, int row) {
+        return false;
```
Comment (TakenouchiTR) [inline code/model/sudoku/SudokuPuzzle.java:103]: Should return whether a number is locked at a specified `row` and `column`; 

`return this.numberLocks[row][column];`.

File: code/model/sudoku/SudokuPuzzle.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * Gets the numbers
+     * 
+     * @return this.numbers
+     */
+    public int[][] getNumber() {
+        return this.numbers;
+    }
```
Comment (TakenouchiTR) [inline code/model/sudoku/SudokuPuzzle.java:35]: I intended for these to be getting the number at a given position, which is why the class diagram specifies a `row` and `column` in the method header.

Recommend changing the name of the current method to be plural (`getNumbers()`), and making a second method that follows what is in the class diagram.

**Review by TakenouchiTR — CHANGES_REQUESTED** [review]

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    private int[][] numbers;
+    private boolean[][] numberLocks;
+    private int[][] defaultAnswer;
+    private int selectedNumber;
+    private Stack<SudokuMove> moveHistory;
+
+    /**
+     * Default Constructor
+     * @precondition none
+     * @postcondition this.numbers = new int[9][9] && this.numberLocks = new boolean[9][9] && this.defaultAnswer = new int[9][9] && this.selectedNumber = -1 && this.moveHistory = new Stack<SudokuMove>()
```
Comment (TakenouchiTR) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:22]: `=` should be `==`
ie: `this.numberLocks = new boolean[9][9]` => `this.numberLocks == new boolean[9][9]`

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        this.defaultAnswer = new int[9][9];
+        this.selectedNumber = -1;
+        this.moveHistory = new Stack<SudokuMove>();
+    }
+
+    /**
+     * Gets the numbers
+     * @precondition none
+     * @postcondition none
+     * @return this.numbers
```
Comment (TakenouchiTR) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:36]: We can't see what a private field is from the outside. The return should not mention it, and should tell the developer what the return value is in words.

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+     */
+    public int[][] getNumbers() {
+        return this.numbers;
+    }
+
+    /**
+     * Gets the move history
+     * @precondition none
+     * @postcondition none
+     * @return this.moveHistory
```
Comment (TakenouchiTR) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:46]: We can't see what a private field is from the outside. The return should not mention it, and should tell the developer what the return value is in words.

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        if (row < 0) {
+            throw new IndexOutOfBoundsException("row can not be less than 0");
+        }
+    }
+
+    /**
+     * Gets the number locks
+     * @precondition none
+     * @postcondition none
+     * @return this.numberLocks
```
Comment (TakenouchiTR) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:88]: We can't see what a private field is from the outside. The return should not mention it, and should tell the developer what the return value is in words.

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    /***
+     * sets the lock at the specified point
+     * 
+     * @param lock
+     * @param column
+     * @param row
+     * @precondition column >= 0 && row >= 0
+     * @postcondition this.numberLocks[row][column] = lock
```
Comment (TakenouchiTR) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:126]: See above about `=` and `==`

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+     */
+    public void setNumberLock(boolean lock, int column, int row) {
+        this.checkColumnAndRow(column, row);
+        this.numberLocks[row][column] = lock;
+    }
+
+    /***
+     * sets the selected number
+     * @precondition number >= 0 && number <= 9
+     * @postcondition this.selectedNumber = number
```
Comment (TakenouchiTR) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:136]: See above `=`

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+     * Checks to see if game is comple
+     * @precondition none
+     * @postcondition none
+     * @return true or false depending on state of the game
+     */
+    public boolean isComplete() {
+        return this.numbers.equals(this.defaultAnswer);
+    }
+
+    /***
```
Comment (TakenouchiTR) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:163]: pls fix ***

File: code/src/test/java/habit_mode/test/model/sudoku/sudokuPuzzle/TestConstructor.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+		
+		SudokuPuzzle puzzle = new SudokuPuzzle();
+        int[][] defaultAnswer = new int[9][9];
+        int[][] numbers = new int[9][9];
+        boolean[][] numberLocks = new boolean[9][9];
+        Stack<SudokuMove> moveHistory = new Stack<SudokuMove>();
+		
+		assertAll(
+            ()->{assertEquals(-1, puzzle.getSelectedNumber(), "checking selectedNumber");}
+		);
```
Comment (TakenouchiTR) [inline code/src/test/java/habit_mode/test/model/sudoku/sudokuPuzzle/TestConstructor.java:25]: Inconsistent tabbing

**Review by TakenouchiTR — CHANGES_REQUESTED** [review]
Please make sure not to add .~ files.

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java (hunk 10 above)
Comment (bfreder1) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:88]: Needs new line

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+     * @return selected number
+     */
+    public int getSelectedNumber() {
+        return this.selectedNumber;
+    }
+
+    /**
+     * Gets the number at specified location
+     * 
+     * @param column the column
```
Comment (bfreder1) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:65]: Re-order this doc

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java (hunk 9 above)
Comment (bfreder1) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:46]: Needs new line

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java (hunk 8 above)
Comment (bfreder1) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:36]: Needs some new lines

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java (hunk 7 above)
Comment (bfreder1) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:22]: Needs new line

File: code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+        this.numbers[row][column] = value;
+    }
+
+    /**
+     * Sets the lock at the specified point.
+     * 
+     * @precondition column >= 0 && row >= 0
+     * @postcondition this.numberLocks[row][column] == lock
+     * 
+     * @param lock
```
Comment (Steven-Carriger) [inline code/src/main/java/habit_mode/model/sudoku/SudokuPuzzle.java:139]: same as before

**Review by Steven-Carriger — CHANGES_REQUESTED** [review]
Overall, everything looks fine. Just a couple documentation issues that I have noticed.

**Review by Steven-Carriger — APPROVED** [review]
All changes have been implemented.

**Pull request merged**